### DISCUSSION TITLE
tutorials/lift-off-part4/07-testing-a-mutation-in-the-explorer

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#80] [course feedback]
# tutorials/lift-off-part4/07-testing-a-mutation-in-the-explorer

https://www.apollographql.com/tutorials/lift-off-part4/07-testing-a-mutation-in-the-explorer

**Comment by omriluz** [comment]
Love how separated each part of the lifecycle of the request is, I am sure it leads to much easier debugging

**Comment by xavier643** [comment]
Where does error.extensions.response get it's information from though? I originally answer the last question that the message came from the REST API because I though ultimately that is where it came from. Isn't that where error.extensions gets it's information?

**Reply by mabuyo** [reply]
Hi @xavier643, you're referring to the last multiple choice quiz in the lesson? Yes, the `error.extensions.response` value is based on what the REST API returns. We'll reword the answers to be a bit more clear. Thanks!

**Reply by xavier643** [reply]
I appreciate that but for me the question is where does it ultimately come from? Is apollo somehow creating it or does the rest api need to be specifically setup to report that? If it's up to the rest api then is it possible that the rest api actually fails to send back a response that is useful or maybe even null?

**Reply by mabuyo** [reply]
Ah, thanks for clarifying! In this case, it's the  `RESTDataSource` class populating these values. You can take a look at the source code here: https://github.com/apollographql/datasource-rest/blob/f1a43237a74a3ff1d8b40b5fa84366ecd3d473d0/src/RESTDataSource.ts#L388

You can see the `extensions.reponse` object is populated with values from the REST API HTTP response.

```ts
return new GraphQLError(`${response.status}: ${response.statusText}`, {
      extensions: {
        ...(code && { code }),
        response: {
          url: response.url,
          status: response.status,
          statusText: response.statusText,
          body: parsedBody,
        },
      },
    });
```

This assumes that the REST API returns those values it's setting but I believe it is a common convention already for the response to contain those. Does this help?

**Reply by xavier643** [reply]
Yes that helps. Thank you for this response. Sorry if this distracted from the purpose of the comment section but I was really curious and having a hard time finding the answer to my question on google for some reason.

**Reply by mabuyo** [reply]
No worries, this is a great place to ask questions! :) I didn't know off the top of my head and had to hunt for the answer as well. Feel free to drop other comments if you run into any confusion.
